src/style-spec/function/compile.js

```diff
@@ -0,0 +1,155 @@
+'use strict';
+// @flow
+
+const assert = require('assert');
+
+module.exports = compileExpression;
+
+const expressions = require('./expressions');
+const parseExpression = require('./parse');
+const typecheck = require('./type_check');
+const evaluationContext = require('./evaluation_context');
+
+/*::
+ import type { PrimitiveType, TypeName, VariantType, VectorType, ArrayType, AnyArrayType, NArgs, LambdaType, Type } from './types.js';
+
+ import type { TypeError, TypedLambdaExpression, TypedLiteralExpression, TypedExpression } from './type_check.js';
+
+ import type { ExpressionName } from './expressions.js';
+
+ export type CompiledExpression = {|
+     result: 'success',
+     js: string,
+     type: Type,
+     isFeatureConstant: boolean,
+     isZoomConstant: boolean,
+     expression: TypedExpression,
+     function?: any
+ |}
+
+
+ type CompileError = {
+     error: string,
+     key: string
+ }
+
+ type CompileErrors = {|
+     result: 'error',
+     errors: Array<CompileError>
+ |}
+ */
+
+/**
+ *
+ * Given a style function expression object, returns:
+ * ```
+ * {
+ *   result: 'success',
+ *   isFeatureConstant: boolean,
+ *   isZoomConstant: boolean,
+ *   js: string,
+ *   function: Function
+ * }
+ * ```
+ * or else
+ *
+ * ```
+ * {
+ *   result: 'error',
+ *   errors: Array<CompileError>
+ * }
+ * ```
+ *
+ * @private
+ */
+function compileExpression(expr: any) {
```

Comment:
Looks like it's possible to narrow `any` to `mixed` here without too much trouble.